src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1511 @@
+        refs = RefSet.newEmpty();
+        int depthPrev = depth;
+        depth = 0;
+        callStack.push(site);
+        try {
+
+            // Add initial references from method receiver
+            refs.addAll(receiverRefs);
+
+            // Add initial references from parameters
+            refs.addAll(paramRefs);
+
+            // Stop trivial cases here
+            if (refs.isEmpty())
+                return;
+
+            // Stop infinite recursion here
+            Pair<JCTree, RefSet<Ref>> invocation = Pair.of(site, refs.clone());
+            if (!invocations.add(invocation))
+                return;
+
+            // Scan method body to "execute" it
+            try {
+                scan(method.body);
+            } finally {
+                invocations.remove(invocation);
+            }
+
+            // "Return" any references from method return value
+            refs.mapInto(refsPrev, ReturnRef.class, direct -> new ExprRef(depthPrev, direct));
+        } finally {
+            callStack.pop();
+            depth = depthPrev;
+            refs = refsPrev;
+            methodClass = methodClassPrev;
+        }
+    }
+
+    // Handle invocation of an unknown or overridable method or constructor
+    private void invokeUnknown(JCTree invoke, List<JCExpression> args, RefSet<?> receiverRefs) {
+
+        // Detect leak via receiver
+        if (!receiverRefs.isEmpty())
+            leakAt(invoke);
+
+        // Detect leaks via method parameters
+        for (JCExpression arg : args) {
+            scan(arg);
+            if (refs.discardExprs(depth))
+                leakAt(arg);
+        }
+    }
+
+//
+// Visitor methods - new Foo()
+//
+
+    @Override
+    public void visitNewClass(JCNewClass tree) {
+        MethodInfo methodInfo = methodMap.get(tree.constructor);
+        if (methodInfo != null && methodInfo.isInvokable())
+            invokeInvokable(tree, tree.args, outerThisRefs(tree.encl, tree.clazz.type), methodInfo);
+        else
+            invokeUnknown(tree, tree.args, outerThisRefs(tree.encl, tree.clazz.type));
+    }
+
+    // Determine 'this' references passed to a constructor via the outer 'this' instance
+    private RefSet<OuterRef> outerThisRefs(JCExpression explicitOuterThis, Type type) {
+        RefSet<OuterRef> outerRefs = RefSet.newEmpty();
+        if (explicitOuterThis != null) {
+            scan(explicitOuterThis);
+            refs.removeExprs(depth, direct -> outerRefs.add(new OuterRef(direct)));
+        } else if (type.tsym != methodClass.sym && type.tsym.isEnclosedBy(methodClass.sym)) {
+            refs.mapInto(outerRefs, ThisRef.class, OuterRef::new);
+        }
+        return outerRefs;
+    }
+
+//
+// Visitor methods - Codey Bits
+//
+
+    @Override
+    public void visitBlock(JCBlock tree) {
+        visitScoped(tree, false, super::visitBlock);
+        Assert.check(checkInvariants(true, false));
+    }
+
+    @Override
+    public void visitDoLoop(JCDoWhileLoop tree) {
```

Comment:
I was thinking, code can also loop using labels and `break` / `continue`, not something we need to cover as part of this prototype but could be a future TODO that we can document

Reply:
Hah - I didn't think of that. But actually I don't think we would miss anything (see if you agree).

The code "executes" every loop, in a sort-of simulation, adding references until the set of references converges. Moreover, that reference set is "append only" while this is happening.

Therefore, during actual execution, a break or continue may cause less code to be executed than in our simulation, but never more code than our simulation. So during actual execution it might be that fewer actual 'this' references are created, but never more.

Therefore, this effect might cause false positives (which of course we already have with loops and code in general because we take all possible branches), but never false negatives.

Reply:
yep I agree